### ISSUE TITLE
Crash in NIO

### ISSUE THREAD

**Issue by bridger** [IBM-Swift/Kitura-WebSocket-NIO#35]
Today I tried deploying the NIO websockets to production. Unfortunately it crashed quickly twice and I had to rollback to a previous version. Even worse, the crash report doesn't have much stack trace (any tips on how to get better diagnostics are much appreciated!) so I don't have a repro.

The error in my logs is this:
```
20:03:57
Fatal error: tried to decode as type HTTPPart<HTTPRequestHead, ByteBuffer> but found IOData with contents ioData(IOData { ByteBuffer { readerIndex: 0, writerIndex: 893, readableBytes: 893, capacity: 1024, slice: _ByteBufferSlice { 0..<1024 }, storage: 0x00007f5150118310 (1024 bytes) } }): file /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/NIOAny.swift, line 200

Illegal instruction
```

I wasn't able to repro this when running with Docker locally. I can try to get a staging environment that is more similar to production, but this is running in Amazon ECS and I'm not sure how to get a better crash report in Amazon CloudWatch.

**Comment by bridger** [comment]
I'm also not positive this is in Websocket. This might be in a different part of the Kitura stack. My app also has very simple REST endpoints.

**Comment by bridger** [comment]
I also got this error logged many times before the crash

```
[2019-07-06T20:07:35.511Z] [ERROR] [WebSocketConnection.swift:81 send(message:asBinary:)] ChannelHandlerContext unavailable
```

**Comment by pushkarnk** [comment]
Thanks for reporting this issue.

First, an explanation on the error messages:

While fixing a memory leak #32 we made `WebSocketConnection` have only a weak reference to `ChannelHandlerContext`. The latter (and also a `NIO Channel`) can be thought of as an abstraction that, somewhere down the line, refers to an active TCP connection. Once the connection is closed, the `Channel` and hence the `ChannelHandlerContext` cease to exist. The application is probably trying to write to a closed socket. We must elide those error messages based on the reasoning we used in a [similar issue](https://github.com/IBM-Swift/Kitura-NIO/pull/186) in Kitura-NIO.

Why did we start seeing these messages only from release `2.1.0`? Before #32 we held the `ChannelHandlerContext` as a strong reference in `WebSocketConnection`. This created a retain cycle between the two. Even if the underlying channel got closed `WebSocketConnection` which is a channel handler kept it alive. This also caused a memory leak.  A channel handler, if necessary, can only hold a weak reference to the underlying channel.

I'll do a PR to remove the error messages.

**Comment by pushkarnk** [comment]
Now, regarding the crash.

The error message means some data that isn't a valid HTTP request has entered the ChannelPipeline and was stopped at the HTTPDecoder. Whenever we do a WebSocket upgrade, `NIOWebSocket` removes all of its HTTP-related handlers from the pipeline, installing the `WebSocketDecoder` and `WebSocketEncoder`. Subsequently, `Kitura-NIO` also removes its HTTP handler. Only then do we return a HTTP code 101 to the client. Subsequent messages can then be only WebSocket frames.

For now, to me, this error means we've either received a corrupt HTTP header or under some circumstances we've failed to remove the HTTPDecoder from the pipeline before the `WebSocketFrame`s start flowing in.

A repro would really help. I'll anyway continue to go through the code to understand what can go wrong.

**Comment by pushkarnk** [comment]
About your diagnostics question: this issue can be a great use case for @ianpartridge 's [crash backtraces](https://github.com/ianpartridge/swift-backtrace) library. We can help you get it configured in your production environment.

**Comment by pushkarnk** [comment]
PR to stop logging errors on detecting a closed channel: https://github.com/IBM-Swift/Kitura-WebSocket-NIO/pull/36

**Comment by bridger** [comment]
I'll try out the crash backtraces library. I'm moving apartments beginning of next week, so I might not have time to get it implemented for a little while.

For the other error message, I see there is an `active` ivar in the WebsocketConnection that looks like it was meant to prevent messages on closed connections, but it is never set to false. https://github.com/IBM-Swift/Kitura-WebSocket-NIO/blob/b1e4d808c7711eb388a1ad8dd0aba244f3876c4e/Sources/KituraWebSocket/WebSocketConnection.swift#L79

I'll check my application code for cases where I am not cleaning up closed connections.

**Comment by bridger** [comment]
I was able to repro the crash in my staging environment, but the steps were unclear. I got an iPad and a few tabs in my browser connected. Things were working fine. I closed and reopened the browser tabs and was able to get the crash.

This is the stack trace:
```
0x00000000000c5cc1: function signature specialization <Arg[0] = Dead> of closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install() -> () at /root/Scribble-Server/.build/checkouts/swift-backtrace/Sources/Backtrace/Backtrace.swift:12
0x00000000000c56f9: @objc closure #1 (Swift.Int32) -> () in static Backtrace.Backtrace.install() -> () at /root/Scribble-Server/<compiler-generated>:?
0x0000000000010330: __restore_rt at ??:?
0x0000000000321f0a: function signature specialization <Arg[0] = Exploded, Arg[1] = Exploded> of Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never at crtstuff.c:?
0x0000000000148a39: Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never at ??:?
0x00000000003c856f: function signature specialization <Arg[0] = Dead> of generic specialization <NIOHTTP1.HTTPPart<NIOHTTP1.HTTPRequestHead, NIO.ByteBuffer>> of NIO.NIOAny.forceAsOther<A>(type: A.Type) -> A at /root/Scribble-Server/<compiler-generated>:?
0x00000000003c85cb: function signature specialization <Arg[0] = Dead> of generic specialization <NIOHTTP1.HTTPPart<NIOHTTP1.HTTPRequestHead, NIO.ByteBuffer>> of NIO.NIOAny.forceAs<A>(type: A.Type) -> A at /root/Scribble-Server/<compiler-generated>:?
0x00000000004f4e02: generic specialization <NIOHTTP1.HTTPPart<NIOHTTP1.HTTPRequestHead, NIO.ByteBuffer>> of NIO.NIOAny.forceAs<A>(type: A.Type) -> A at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) generic specialization <NIOHTTP1.HTTPServerPipelineHandler> of (extension in NIO):NIO.ChannelInboundHandler.unwrapInboundIn(NIO.NIOAny) -> A.InboundIn at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) NIOHTTP1.HTTPServerPipelineHandler.(deliverOneMessage in _60DBF281C59552393B20BFC998B34533)(context: NIO.ChannelHandlerContext, data: NIO.NIOAny) -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIOHTTP1/HTTPServerPipelineHandler.swift:192
0x00000000004f4c5e: NIOHTTP1.HTTPServerPipelineHandler.channelRead(context: NIO.ChannelHandlerContext, data: NIO.NIOAny) -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIOHTTP1/HTTPServerPipelineHandler.swift:187
0x00000000004f6259: protocol witness for NIO._ChannelInboundHandler.channelRead(context: NIO.ChannelHandlerContext, data: NIO.NIOAny) -> () in conformance NIOHTTP1.HTTPServerPipelineHandler : NIO._ChannelInboundHandler in NIOHTTP1 at /root/Scribble-Server/<compiler-generated>:?
0x000000000044a14b: NIO.ChannelHandlerContext.(invokeChannelRead in _EEC863903996E9F191EBAFEB0FB0DFDD)(NIO.NIOAny) -> () at /root/Scribble-Server/<compiler-generated>:?
0x0000000000442dc9: NIO.ChannelHandlerContext.fireChannelRead(NIO.NIOAny) -> () at /root/Scribble-Server/<compiler-generated>:?
0x00000000004e76b5: NIOHTTP1.HTTPDecoder.decodeLast(context: NIO.ChannelHandlerContext, buffer: inout NIO.ByteBuffer, seenEOF: Swift.Bool) throws -> NIO.DecodingState at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIOHTTP1/HTTPDecoder.swift:?
0x00000000004e7800: protocol witness for NIO.ByteToMessageDecoder.decodeLast(context: NIO.ChannelHandlerContext, buffer: inout NIO.ByteBuffer, seenEOF: Swift.Bool) throws -> NIO.DecodingState in conformance NIOHTTP1.HTTPDecoder<A, B> : NIO.ByteToMessageDecoder in NIOHTTP1 at /root/Scribble-Server/<compiler-generated>:?
0x00000000004610ac: closure #1 (inout A, inout NIO.ByteBuffer) throws -> NIO.DecodingState in NIO.ByteToMessageHandler.(decodeLoop in _F2A740607FEBA425AF6F8C49A24C0AD7)(context: NIO.ChannelHandlerContext, decodeMode: NIO.ByteToMessageHandler<A>.(DecodeMode in _F2A740607FEBA425AF6F8C49A24C0AD7)) throws -> NIO.(B2MDBuffer in _F2A740607FEBA425AF6F8C49A24C0AD7).BufferProcessingResult at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Codec.swift:?
0x0000000000460da7: NIO.ByteToMessageHandler.(withNextBuffer in _F2A740607FEBA425AF6F8C49A24C0AD7)(allowEmptyBuffer: Swift.Bool, _: (inout A, inout NIO.ByteBuffer) throws -> NIO.DecodingState) throws -> NIO.(B2MDBuffer in _F2A740607FEBA425AF6F8C49A24C0AD7).BufferProcessingResult at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Codec.swift:507
 (inlined by) NIO.ByteToMessageHandler.(decodeLoop in _F2A740607FEBA425AF6F8C49A24C0AD7)(context: NIO.ChannelHandlerContext, decodeMode: NIO.ByteToMessageHandler<A>.(DecodeMode in _F2A740607FEBA425AF6F8C49A24C0AD7)) throws -> NIO.(B2MDBuffer in _F2A740607FEBA425AF6F8C49A24C0AD7).BufferProcessingResult at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Codec.swift:541
0x00000000004607fd: NIO.ByteToMessageHandler.(processLeftovers in _F2A740607FEBA425AF6F8C49A24C0AD7)(context: NIO.ChannelHandlerContext) -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Codec.swift:518
0x00000000004618c0: NIO.ChannelHandlerContext.fireChannelInactive() -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/ChannelPipeline.swift:1136
 (inlined by) NIO.ByteToMessageHandler.channelInactive(context: NIO.ChannelHandlerContext) -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Codec.swift:646
0x0000000000461999: protocol witness for NIO._ChannelInboundHandler.channelInactive(context: NIO.ChannelHandlerContext) -> () in conformance NIO.ByteToMessageHandler<A> : NIO._ChannelInboundHandler in NIO at /root/Scribble-Server/<compiler-generated>:?
0x000000000044a001: NIO.ChannelHandlerContext.(invokeChannelInactive in _EEC863903996E9F191EBAFEB0FB0DFDD)() -> () at /root/Scribble-Server/<compiler-generated>:?
0x000000000044a033: NIO.ChannelHandlerContext.(invokeChannelInactive in _EEC863903996E9F191EBAFEB0FB0DFDD)() -> () at /root/Scribble-Server/<compiler-generated>:?
0x000000000041c9ec: closure #6 (NIO.EventLoopPromise<()>?, NIO.ChannelPipeline) -> () in NIO.(SocketChannelLifecycleManager in _9C3D3FC8F9F3D83A05D3F6063F6AE958).moveState(event: NIO.(SocketChannelLifecycleManager in _9C3D3FC8F9F3D83A05D3F6063F6AE958).Event) -> (NIO.EventLoopPromise<()>?, NIO.ChannelPipeline) -> () at /root/Scribble-Server/<compiler-generated>:?
0x0000000000420b55: generic specialization <NIO.Socket> of NIO.BaseSocketChannel.close0(error: Swift.Error, mode: NIO.CloseMode, promise: NIO.EventLoopPromise<()>?) -> () at /root/Scribble-Server/<compiler-generated>:?
0x00000000004cbdc8: NIO.SocketChannel.close0(error: Swift.Error, mode: NIO.CloseMode, promise: NIO.EventLoopPromise<()>?) -> () at /root/Scribble-Server/<compiler-generated>:?
0x000000000047a2ce: function signature specialization <Arg[2] = Dead> of generic specialization <NIO.SocketChannel> of NIO.SelectableEventLoop.(handleEvent in _D5D78C61B22284700B9BD1ACFBC25157)<A where A: NIO.SelectableChannel>(_: NIO.SelectorEventSet, channel: A) -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/BaseSocketChannel.swift:?
0x00000000004c3df1: generic specialization <NIO.SocketChannel> of NIO.SelectableEventLoop.(handleEvent in _D5D78C61B22284700B9BD1ACFBC25157)<A where A: NIO.SelectableChannel>(_: NIO.SelectorEventSet, channel: A) -> () at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) closure #1 (NIO.SelectorEvent<NIO.NIORegistration>) -> () in closure #1 () throws -> () in NIO.SelectableEventLoop.run() throws -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/EventLoop.swift:833
 (inlined by) reabstraction thunk helper from @callee_guaranteed (@guaranteed NIO.SelectorEvent<NIO.NIORegistration>) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed NIO.SelectorEvent<NIO.NIORegistration>) -> (@error @owned Swift.Error) at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) function signature specialization <Arg[1] = [Closure Propagated : closure #1 (NIO.SelectorEvent<NIO.NIORegistration>) -> () in closure #1 () throws -> () in NIO.SelectableEventLoop.run() throws -> (), Argument Types : [NIO.SelectableEventLoop]> of generic specialization <NIO.NIORegistration> of NIO.Selector.whenReady(strategy: NIO.SelectorStrategy, _: (NIO.SelectorEvent<A>) throws -> ()) throws -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/Selector.swift:537
0x000000000047269f: closure #1 () throws -> () in NIO.SelectableEventLoop.run() throws -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/EventLoop.swift:828
 (inlined by) reabstraction thunk helper from @callee_guaranteed () -> (@error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error) at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) generic specialization <()> of NIO.(withAutoReleasePool in _D5D78C61B22284700B9BD1ACFBC25157)<A>(() throws -> A) throws -> A at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/EventLoop.swift:584
 (inlined by) NIO.SelectableEventLoop.run() throws -> () at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/EventLoop.swift:827
0x000000000047479e: $defer #1 () -> () in closure #1 (NIO.NIOThread) -> () in static NIO.MultiThreadedEventLoopGroup.(setupThreadAndEventLoop in _D5D78C61B22284700B9BD1ACFBC25157)(name: Swift.String, initializer: (NIO.NIOThread) -> ()) -> NIO.SelectableEventLoop at /root/Scribble-Server/<compiler-generated>:?
 (inlined by) closure #1 (NIO.NIOThread) -> () in static NIO.MultiThreadedEventLoopGroup.(setupThreadAndEventLoop in _D5D78C61B22284700B9BD1ACFBC25157)(name: Swift.String, initializer: (NIO.NIOThread) -> ()) -> NIO.SelectableEventLoop at /root/Scribble-Server/.build/checkouts/swift-nio/Sources/NIO/EventLoop.swift:1032
0x000000000047d4b9: partial apply forwarder for closure #1 (NIO.NIOThread) -> () in static NIO.MultiThreadedEventLoopGroup.(setupThreadAndEventLoop in _D5D78C61B22284700B9BD1ACFBC25157)(name: Swift.String, initializer: (NIO.NIOThread) -> ()) -> NIO.SelectableEventLoop at /root/Scribble-Server/<compiler-generated>:?
0x000000000047d4d4: partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed NIO.NIOThread) -> () to @escaping @callee_guaranteed (@in_guaranteed NIO.NIOThread) -> (@out ()) at /root/Scribble-Server/<compiler-generated>:?
0x00000000004d976c: closure #1 (Swift.UnsafeMutableRawPointer?) -> Swift.UnsafeMutableRawPointer? in static NIO.NIOThread.spawnAndRun(name: Swift.String?, body: (NIO.NIOThread) -> ()) -> () at /root/Scribble-Server/<compiler-generated>:?
0x00000000004d97b9: @objc closure #1 (Swift.UnsafeMutableRawPointer?) -> Swift.UnsafeMutableRawPointer? in static NIO.NIOThread.spawnAndRun(name: Swift.String?, body: (NIO.NIOThread) -> ()) -> () at /root/Scribble-Server/<compiler-generated>:?
0x0000000000008184: start_thread at ??:?
0x000000000000000c: ?? ??:0
```

**Comment by bridger** [comment]
I can't find any pattern in the repro. It seems to be just from opening and closing about a dozen connections. I'll just keep cycling through browser tabs doing the same thing (connecting to the server, sending a few messages, close the tab) and then after a few minutes it will crash. I've been testing mostly with Firefox but I think it crashed with Safari too.

**Comment by ianpartridge** [comment]
Thanks for the info @bridger. @pushkarnk is out today but will be back tomorrow and we'll take a look ASAP.

**Comment by pushkarnk** [comment]
@bridger Thanks for the backtrace.

You said the server crashed after you closed the browser tabs and opened new tabs. I have questions:
1. Did the browser tabs have active WebSocket connections?
2. Did you see the soon crash after you closed the tabs or after you opened the new tabs, trying to create new WS connections?

The backtrace indicates that this crash happened on an HTTP connection, not on a WS one. A socket close event was being propagated through the HTTP server pipeline through a chain of  `channelInactive()` calls.

On a `channelInactive` event the `ByteToMessageHandler<HTTPDecoder>` tries to `decodeLeftOvers()` and passes contents of the ByteBuffer to the next handler `HTTPServerPipelineHandler` which tries to unwrap this data as an HTTPRequest part, which the data is not, evidently. Hence the fatalError.

A couple of inferences:
1. The closing of browser tabs may have caused the socket close event.
2. Presence of HTTP handlers in the pipeline indicates this didn't occur on a WS connection. 
3. We don't know what data the socket buffer held when the connection was abruptly closed. It could have been a partial  WS upgrade request (the size is 893 bytes?).

**Comment by pushkarnk** [comment]
The only failure hypothesis I have is this: The browser tabs were in the process of making a WS connection, when they were abruptly closed. Possibly a part of the request was sent to the server before the connections were closed. On the server, we tried to process the partially received data as an `HTTPRequestPart` and fail.

**Comment by pushkarnk** [comment]
@weissi Need your support here. Can you please take a look at the backtrace [here](https://github.com/IBM-Swift/Kitura-WebSocket-NIO/issues/35#issuecomment-509472213) and help us understand what could be going wrong?

**Comment by pushkarnk** [comment]
The only failure hypothesis I have : One of the browser tabs was in the process of making a WS connection, when it was abruptly closed. Possibly, a part of the request was read by server before the connection was closed. On the server, we tried to process the partially received data as an `HTTPRequestPart` and failed.

**Comment by weissi** [comment]
@pushkarnk / @bridger Thanks for the investigation, and aplogies. What's happening here is a NIO bug. What seems to be going on is that the client starts an HTTP upgrade but before it's finished the connection is closed.

What the code in NIO that forwards the left-overs is trying to achieve is:
- After the upgrade is complete, forward the left-over bytes to the next handler

That makes sense because some bytes at the end of the HTTP upgrade request might be the first couple of WebSocket bytes. The thing that's slightly odd in NIO totday is that it is using the invocation of `decodeLast` the signal for the upgrade being complete. Whilst that works (because on completion of the upgrade you must (and do) remove the handler which triggers `decodeLast`). However, and that's what we missed is that `decodeLast` isn't only invoked when the handler is removed but also when the connection is closed... So in the unfortunate case that we do have left-overs (ie. the upgrade is in progress) and the connection gets torn down, we send stuff through the pipeline which will invariably force the next handler to cause a crash.

Long story short, we'll fix that and make the left-overs forwarding only happen iff the HTTPHandler gets removed (and not when the connection gets torn down (which is a terrible idea because it's not under control of the NIO user)).

I'm working on a patch, the quickest patch is literally just a `if !seenEOF { ... }` but I need to write tests etc :)

**Comment by weissi** [comment]
@pushkarnk it's merged to master and will be part of NIO 2.4.0 which we will release very soon, release tools to validate the release are running. @pushkarnk / @bridger would you be able to verify the fix before the 2.4.0 release? I'm very confident that this is the issue but having it verified would be even better :)

**Comment by bridger** [comment]
Thank you everyone for investigating and fixing the bug! I just deployed to my staging environment and so far it seems good. I'll keep trying to repro but I'm hopeful!

I tried to repro by refreshing the browser a lot so that it will continuously make new connections and try to get interrupted. I didn't get the crash, but I see my application code believes there are lots of open connections. That may be a bug in the way I handle the connectionClosed handler, or maybe I'm not getting the connectionClosed callback.

**Comment by pushkarnk** [comment]
Thanks @weissi on your investigation and the quick fix! Appreciate it :)

**Comment by pushkarnk** [comment]
> Thank you everyone for investigating and fixing the bug! I just deployed to my staging environment and so far it seems good. I'll keep trying to repro but I'm hopeful!
> 
> I tried to repro by refreshing the browser a lot so that it will continuously make new connections and try to get interrupted. I didn't get the crash, but I see my application code believes there are lots of open connections. That may be a bug in the way I handle the connectionClosed handler, or maybe I'm not getting the connectionClosed callback.

@bridger Thanks for testing and confirming.
Regarding the other issue,  I tested a very simple scenario of handful browser tabs connecting to a ws server and disconnecting. I could see the `disconnected(connection:reason)` callback being invoked. I'll wait for more details what you observed.

**Comment by weissi** [comment]
Thanks so much @bridger for the validation. We're going to release 2.4.0 today. All the release tools have run overnight and came back green.

**Comment by weissi** [comment]
@bridger / @pushkarnk NIO 2.4.0 released, so a `from: "2.4.0"` in your `Package.swift` should make sure we don't crash you anymore. Apologies again for the breakage.
